File: apps/patter-voice-sdk/libraries/typescript/src/providers/cartesia-stt.ts
/**
 * Cartesia STT (ink-whisper) adapter for the Patter SDK pipeline mode.
 *
 * Implements a `DeepgramSTT`-shaped provider using Cartesia's streaming
 * WebSocket API. Pure `ws` transport — does NOT depend on the vendor SDK.
 */

import WebSocket from 'ws';
import { getLogger } from '../logger';

/** Patter-normalised transcript event emitted by {@link CartesiaSTT}. */
export interface Transcript {
  readonly text: string;
  readonly isFinal: boolean;
  readonly confidence: number;
}

type TranscriptCallback = (transcript: Transcript) => void;

/** Known Cartesia STT models. */
export const CartesiaSTTModel = {
  INK_WHISPER: 'ink-whisper',
} as const;
export type CartesiaSTTModel = (typeof CartesiaSTTModel)[keyof typeof CartesiaSTTModel];

/** Audio encodings accepted by Cartesia's STT websocket endpoint. */
export const CartesiaSTTEncoding = {
  PCM_S16LE: 'pcm_s16le',
} as const;
export type CartesiaSTTEncoding = (typeof CartesiaSTTEncoding)[keyof typeof CartesiaSTTEncoding];

/** Common PCM sample rates accepted by Cartesia STT. */
export const CartesiaSTTSampleRate = {
  HZ_8000: 8000,
  HZ_16000: 16000,
  HZ_24000: 24000,
  HZ_44100: 44100,
  HZ_48000: 48000,
} as const;
export type CartesiaSTTSampleRate = (typeof CartesiaSTTSampleRate)[keyof typeof CartesiaSTTSampleRate];

/** Cartesia STT server event `type` values. */
export const CartesiaSTTServerEvent = {
  TRANSCRIPT: 'transcript',
  FLUSH_DONE: 'flush_done',
  DONE: 'done',
  ERROR: 'error',
} as const;
export type CartesiaSTTServerEvent = (typeof CartesiaSTTServerEvent)[keyof typeof CartesiaSTTServerEvent];

/** Cartesia STT client-side text frames. */
export const CartesiaSTTClientFrame = {
  FINALIZE: 'finalize',
} as const;
export type CartesiaSTTClientFrame = (typeof CartesiaSTTClientFrame)[keyof typeof CartesiaSTTClientFrame];

/** Cartesia STT currently only accepts 16-bit PCM little-endian. */
/** Legacy encoding alias kept for callers using the bare string form. */
export type CartesiaEncoding = 'pcm_s16le';

/** Constructor options for {@link CartesiaSTT}. */
export interface CartesiaSTTOptions {
  /** Cartesia STT model. Currently only `"ink-whisper"`. */
  readonly model?: CartesiaSTTModel | string;
  /** BCP-47 language code. */
  readonly language?: string;
  /** PCM encoding; Cartesia only supports `pcm_s16le`. */
  readonly encoding?: CartesiaSTTEncoding | CartesiaEncoding;
  /** Sample rate in Hz. Cartesia accepts 8000, 16000, 24000, 44100, 48000. */
  readonly sampleRate?: CartesiaSTTSampleRate | number;
  /** Override base URL (HTTP or WS). Defaults to Cartesia prod. */
  readonly baseUrl?: string;
}

const DEFAULT_BASE_URL = 'https://api.cartesia.ai';
const API_VERSION = '2025-04-16';
const USER_AGENT = 'Patter/1.0';
const KEEPALIVE_INTERVAL_MS = 30000;
const CONNECT_TIMEOUT_MS = 10000;

interface CartesiaEvent {
  readonly type?: string;
  readonly text?: string;
  readonly is_final?: boolean;
  readonly probability?: number;
  readonly request_id?: string;
  readonly message?: string;
}

/** Streaming STT adapter for Cartesia's ink-whisper WebSocket API. */
export class CartesiaSTT {
  /** Stable pricing/dashboard key — read by stream-handler/metrics. */
  static readonly providerKey = 'cartesia_stt';
  private ws: WebSocket | null = null;
  private callbacks: Set<TranscriptCallback> = new Set();
  private keepaliveTimer: ReturnType<typeof setInterval> | null = null;
  /**
   * Cartesia request id — set from the server transcript events.
   * `null` until the first transcript event arrives (matches Python's `None`).
   */
  public requestId: string | null = null;

  /** Construction args replayed by clone(). */
  private readonly patterCtorArgs: unknown[];

  constructor(
    private readonly apiKey: string,
    private readonly options: CartesiaSTTOptions = {},
  ) {
    this.patterCtorArgs = [apiKey, options];
    if (!apiKey) {
      throw new Error('CartesiaSTT requires a non-empty apiKey');
    }
  }

  /**
   * Open a fresh WebSocket without arming any message / keepalive handlers
   * and without taking ownership on `this.ws`. Returns the OPEN socket so
   * the caller (the prewarm pipeline) can park it for later adoption via
   * `adoptWebSocket`. Bounded by `CONNECT_TIMEOUT_MS`.
   *
   * Billing safety: opening + parking the WS does not stream audio
   * (Cartesia STT bills on streamed audio seconds), so no charge is
   * incurred. Close the returned WS yourself if it is never adopted.
   */
  async openParkedConnection(): Promise<WebSocket> {
    const url = this.buildWsUrl();
    const ws = new WebSocket(url, {
      headers: { 'User-Agent': USER_AGENT },
    });
    await new Promise<void>((resolve, reject) => {
      const timer = setTimeout(
        () => reject(new Error('Cartesia STT park connect timeout')),
        CONNECT_TIMEOUT_MS,
      );
      ws.once('open', () => {
        clearTimeout(timer);
        resolve();
      });
      ws.once('error', (err: Error) => {
        clearTimeout(timer);
        // Sanitize before re-throwing: raw `ws` handshake errors can include
        // the upgrade URL, which carries the API key as a query-string param.
        reject(new Error(`Cartesia STT park connect failed: ${describeWarmupError(err)}`));
      });
    });
    return ws;
  }

  private buildWsUrl(): string {
    const opts = this.options;
    const rawBase = opts.baseUrl ?? DEFAULT_BASE_URL;
    let base: string;
    if (rawBase.startsWith('http://')) {
      base = `ws://${rawBase.slice('http://'.length)}`;
    } else if (rawBase.startsWith('https://')) {
      base = `wss://${rawBase.slice('https://'.length)}`;
    } else if (rawBase.startsWith('ws://') || rawBase.startsWith('wss://')) {
      base = rawBase;
    } else {
      base = `wss://${rawBase}`;
    }

    const language = opts.language ?? 'en';
    const params = new URLSearchParams({
      model: opts.model ?? CartesiaSTTModel.INK_WHISPER,
      sample_rate: String(opts.sampleRate ?? CartesiaSTTSampleRate.HZ_16000),
      encoding: opts.encoding ?? CartesiaSTTEncoding.PCM_S16LE,
      cartesia_version: API_VERSION,
      api_key: this.apiKey,
      language,
    });
    return `${base}/stt/websocket?${params.toString()}`;
  }

  /**
   * Pre-call WebSocket warmup for the Cartesia STT `/stt/websocket` endpoint.
   *
   * Opens the WS (DNS + TLS + auth handshake), idles ~250 ms so the
   * Cartesia edge keeps session state warm, then closes. By the time
   * `connect()` is invoked at call-pickup the resolver and TLS session
   * are hot — net wire time saving of 200-500 ms.
   *
   * Billing safety: Cartesia STT bills on streamed audio seconds (per
   * https://docs.cartesia.ai/2025-04-16/api-reference/stt/stt). Opening
   * + closing the WebSocket without forwarding audio does not consume
   * billable seconds. Best-effort: failures logged at debug level.
   */
  async warmup(): Promise<void> {
    const url = this.buildWsUrl();
    let ws: WebSocket | null = null;
    try {
      ws = await new Promise<WebSocket>((resolve, reject) => {
        const sock = new WebSocket(url, {
          headers: { 'User-Agent': USER_AGENT },
        });
        const timer = setTimeout(() => {
          try {
            sock.close();
          } catch {
            // ignore
          }
          reject(new Error('Cartesia STT warmup connect timeout'));
        }, 5000);
        sock.once('open', () => {
          clearTimeout(timer);
          resolve(sock);
        });
        sock.once('error', (err: Error) => {
          clearTimeout(timer);
          reject(err);
        });
      });
      // Idle briefly so the provider edge keeps session state warm.
      await new Promise<void>((r) => setTimeout(r, 250));
    } catch (err) {
      // IMPORTANT: ``String(err)`` for a `ws` handshake failure can
      // include the request URL, which carries the API key as a
      // query-string parameter (Cartesia auth pattern). Log only the
      // HTTP status (when present) or the error class name — never the
      // full URL or message.
      getLogger().debug(
        `Cartesia STT warmup failed (best-effort): ${describeWarmupError(err)}`,
      );
    } finally {
      if (ws) {
        try {
          ws.close();
        } catch {
          // ignore
        }
      }
    }
  }

  /** Open the streaming WebSocket and arm message + keepalive handlers. */

  /**
   * Fresh adapter built with this instance's construction arguments —
   * called per call by the stream handler so concurrent calls never share
   * connection state (sockets/queues; cross-call transcript bleed).
   */
  clone(): this {
    const ctor = this.constructor as new (...args: unknown[]) => this;
    return new ctor(...this.patterCtorArgs);
  }

  async connect(): Promise<void> {
    const url = this.buildWsUrl();
    this.ws = new WebSocket(url, {
      headers: { 'User-Agent': USER_AGENT },
    });

    await new Promise<void>((resolve, reject) => {
      const timer = setTimeout(
        () => reject(new Error('Cartesia STT connect timeout')),
        CONNECT_TIMEOUT_MS,
      );
      this.ws!.once('open', () => {
        clearTimeout(timer);
        resolve();
      });
      this.ws!.once('error', (err: Error) => {
        clearTimeout(timer);
        reject(err);
      });
    });

    this.armMessageAndKeepalive();
  }

  /**
   * Adopt a pre-opened, already-OPEN WebSocket produced by the prewarm
   * pipeline (see `Patter.parkProviderConnections`). Skips the fresh
   * `new WebSocket()` + handshake — the WS is already through DNS, TLS
   * and HTTP-101 so audio frames can flow on this turn instead of
   * paying ~150-400 ms of handshake.
   *
   * Caller MUST verify `ws.readyState === OPEN` before calling. If the
   * parked WS died between park and adopt, fall back to `connect()`.
   */
  adoptWebSocket(ws: WebSocket): void {
    this.ws = ws;
    this.armMessageAndKeepalive();
  }

  private armMessageAndKeepalive(): void {
    if (!this.ws) return;
    this.ws.on('message', (raw: WebSocket.RawData) => {
      let event: CartesiaEvent;
      try {
        event = JSON.parse(raw.toString()) as CartesiaEvent;
      } catch {
        return;
      }
      this.handleEvent(event);
    });

    this.keepaliveTimer = setInterval(() => {
      if (this.ws && this.ws.readyState === WebSocket.OPEN) {
        try {
          this.ws.ping();
        } catch {
          // ignore transient ping errors
        }
      }
    }, KEEPALIVE_INTERVAL_MS);
  }

  private handleEvent(event: CartesiaEvent): void {
    const type = event.type;
    if (type === CartesiaSTTServerEvent.TRANSCRIPT) {
      const text = (event.text ?? '').trim();
      const isFinal = Boolean(event.is_final);
      if (!text && !isFinal) return;
      if (event.request_id) {
        this.requestId = event.request_id;
      }
      if (!text) return;
      const confidence = Number(event.probability ?? 1);
      this.emit({ text, isFinal, confidence });
      return;
    }

    if (type === CartesiaSTTServerEvent.ERROR) {
      getLogger().error(`Cartesia STT error: ${event.message ?? 'unknown'}`);
      return;
    }
    // `flush_done` and `done` are informational; no transcript to surface.
  }

  private emit(transcript: Transcript): void {
    for (const cb of this.callbacks) {
      // The registered callback is async (stream-handler's handleTranscript)
      // — a bare call left its rejection unhandled, which kills the Node
      // process. Contain both sync throws and async rejections.
      try {
        Promise.resolve(cb(transcript)).catch((err) =>
          getLogger().error(`STT transcript callback failed: ${String(err)}`),
        );
      } catch (err) {
        getLogger().error(`STT transcript callback threw: ${String(err)}`);
      }
    }
  }

  /** Send a binary PCM16-LE audio chunk to Cartesia for transcription. */
  sendAudio(audio: Buffer): void {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) return;
    this.ws.send(audio);
  }

  /**
   * Force Cartesia to finalise the in-flight utterance immediately.
   *
   * Sends a ``finalize`` text frame on the live WebSocket. Cartesia
   * replies with the final transcript followed by ``flush_done``,
   * bypassing its conservative internal silence heuristic (which can
   * wait 2-7 s on PSTN audio before naturally finalising). Wired
   * into ``StreamHandler`` on the VAD ``speech_end`` event so the
   * SDK's authoritative end-of-speech detection forces an immediate
   * STT finalisation — turning Cartesia's natural-pause endpointing
   * into a deterministic VAD-driven one, parity with the Deepgram
   * fast-path. No-op when the WS isn't open. Parity with Python
   * ``CartesiaSTT.finalize``.
   */
  async finalize(): Promise<void> {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) return;
    await new Promise<void>((resolve) => {
      this.ws!.send(CartesiaSTTClientFrame.FINALIZE, (err) => {
        if (err) {
          getLogger().debug(`Cartesia finalize send failed: ${String(err)}`);
        }
        resolve();
      });
    });
  }

  /** Register a transcript listener. */
  onTranscript(callback: TranscriptCallback): void {
    this.callbacks.add(callback);
  }

  /** Remove a previously registered transcript callback. */
  offTranscript(callback: TranscriptCallback): void {
    this.callbacks.delete(callback);
  }

  /**
   * Synchronous best-effort close. Sends `finalize` and closes the socket
   * without waiting for the server to flush any remaining transcripts.
   *
   * Limitation: any transcript events produced between the `finalize` send
   * and the socket close may be dropped. Callers that need to guarantee all
   * transcripts are delivered should await :meth:`closeAsync` instead.
   */
  close(): void {
    if (this.keepaliveTimer) {
      clearInterval(this.keepaliveTimer);
      this.keepaliveTimer = null;
    }
    if (this.ws) {
      try {
        this.ws.send(CartesiaSTTClientFrame.FINALIZE);
      } catch {
        // ignore
      }
      this.ws.close();
      this.ws = null;
    }
  }

  /**
   * Graceful close that awaits the `finalize` send and the socket closing
   * handshake, matching the Python adapter's behavior. Use this when you
   * need any in-flight transcripts to be flushed before teardown.
   */
  async closeAsync(): Promise<void> {
    if (this.keepaliveTimer) {
      clearInterval(this.keepaliveTimer);
      this.keepaliveTimer = null;
    }
    const ws = this.ws;
    this.ws = null;
    if (!ws) return;

    // Best-effort: send `finalize` so the server flushes any trailing transcripts.
    if (ws.readyState === WebSocket.OPEN) {
      try {
        await new Promise<void>((resolve) => {
          ws.send(CartesiaSTTClientFrame.FINALIZE, (err) => {
            if (err) getLogger().warn(`CartesiaSTT finalize send failed: ${String(err)}`);
            resolve();
          });
        });
      } catch (err) {
        getLogger().warn(`CartesiaSTT finalize error: ${String(err)}`);
      }
    }

    // Wait for the socket to fully close so any final transcript events are
    // delivered through the existing message handler before we return.
    if (ws.readyState === WebSocket.OPEN || ws.readyState === WebSocket.CONNECTING) {
      await new Promise<void>((resolve) => {
        const done = (): void => {
          ws.off('close', done);
          ws.off('error', done);
          resolve();
        };
        ws.once('close', done);
        ws.once('error', done);
        try {
          ws.close();
        } catch {
          // ignore
          resolve();
        }
      });
    }
  }
}

/**
 * Render a warmup error for logging without leaking the request URL.
 *
 * `String(err)` on a `ws` handshake failure can include the upgrade
 * URL, which for Cartesia / AssemblyAI carries the API key as a
 * query-string parameter. This helper extracts only the HTTP status
 * (when present) or the error class name so the API key never lands
 * in logs.
 */
function describeWarmupError(err: unknown): string {
  if (typeof err === 'object' && err !== null) {
    // `ws` handshake failures expose `statusCode` (or `code` on some
    // versions) when the server returned an HTTP error during upgrade.
    const e = err as { statusCode?: number; code?: number; name?: string; constructor?: { name?: string } };
    if (typeof e.statusCode === 'number') return `HTTP ${e.statusCode}`;
    if (typeof e.code === 'number' && e.code >= 100 && e.code < 600) return `HTTP ${e.code}`;
    const ctor = e.constructor?.name;
    if (typeof ctor === 'string' && ctor !== 'Object') return ctor;
    if (typeof e.name === 'string') return e.name;
  }
  // Fallback: log the type, never the full string (which may contain URL).
  return typeof err;
}

File: apps/patter-voice-sdk/libraries/python/getpatter/providers/cartesia_stt.py
"""
Cartesia STT (ink-whisper) adapter for the Patter SDK pipeline mode.

Implements the STTProvider ABC using Cartesia's streaming WebSocket API.
Pure-aiohttp transport — does NOT depend on the vendor SDK.
"""

from __future__ import annotations

import asyncio
import json
import logging
from dataclasses import dataclass
from enum import IntEnum, StrEnum
from typing import ClassVar, AsyncIterator, Literal, Union
from urllib.parse import urlencode

import aiohttp

from getpatter.providers.base import STTProvider, Transcript

logger = logging.getLogger("getpatter")

# Cartesia REST/WS base and protocol constants (port of cartesia/constants.py).
DEFAULT_BASE_URL = "https://api.cartesia.ai"
API_VERSION = "2025-04-16"
USER_AGENT = "Patter/1.0 (provider=Cartesia)"
KEEPALIVE_INTERVAL_SECONDS = 30.0


class CartesiaSTTModel(StrEnum):
    """Known Cartesia STT models."""

    INK_WHISPER = "ink-whisper"


class CartesiaSTTEncoding(StrEnum):
    """Audio encodings accepted by Cartesia's STT websocket endpoint."""

    PCM_S16LE = "pcm_s16le"


class CartesiaSTTSampleRate(IntEnum):
    """Common PCM sample rates accepted by Cartesia STT."""

    HZ_8000 = 8000
    HZ_16000 = 16000
    HZ_24000 = 24000
    HZ_44100 = 44100
    HZ_48000 = 48000


class CartesiaSTTServerEvent(StrEnum):
    """Cartesia STT server event ``type`` values."""

    TRANSCRIPT = "transcript"
    FLUSH_DONE = "flush_done"
    DONE = "done"
    ERROR = "error"


class CartesiaSTTClientFrame(StrEnum):
    """Cartesia STT client-side text frames."""

    FINALIZE = "finalize"


# Backward-compatible Literal aliases — preserved verbatim.
CartesiaEncoding = Literal["pcm_s16le"]
CartesiaModel = Literal["ink-whisper"]


@dataclass
class CartesiaSTTOptions:
    """Configuration for :class:`CartesiaSTT`.

    See https://docs.cartesia.ai/2025-04-16/api-reference/stt/stt
    """

    model: Union[CartesiaSTTModel, str] = CartesiaSTTModel.INK_WHISPER
    language: str = "en"
    encoding: Union[CartesiaSTTEncoding, str] = CartesiaSTTEncoding.PCM_S16LE
    sample_rate: Union[CartesiaSTTSampleRate, int] = CartesiaSTTSampleRate.HZ_16000


class CartesiaSTT(STTProvider):
    """Cartesia STT adapter.

    Streams PCM audio over WebSocket to Cartesia's STT endpoint and yields
    :class:`~getpatter.providers.base.Transcript` objects. Cartesia emits
    interim + final transcripts distinguished by the ``is_final`` boolean
    on the ``transcript`` event.

    Args:
        api_key: Cartesia API key. Required.
        language: BCP-47 language code (``"en"``, ``"es"``, ``"fr"``, ...).
        model: Cartesia STT model; currently only ``"ink-whisper"``.
        encoding: Audio encoding. Only ``"pcm_s16le"`` is supported.
        sample_rate: PCM sample rate in Hz. Cartesia accepts common rates
            (8000, 16000, 24000, 44100, 48000).
        base_url: Override base URL (HTTP or WS). Defaults to Cartesia prod.
        options: Full :class:`CartesiaSTTOptions`; overrides the individual
            kwargs when both are provided.
    """

    #: Stable pricing/dashboard key — read by stream-handler/metrics.
    provider_key: ClassVar[str] = "cartesia_stt"

    def __init__(
        self,
        api_key: str,
        *,
        language: str = "en",
        model: Union[CartesiaSTTModel, str] = CartesiaSTTModel.INK_WHISPER,
        encoding: Union[CartesiaSTTEncoding, str] = CartesiaSTTEncoding.PCM_S16LE,
        sample_rate: Union[CartesiaSTTSampleRate, int] = CartesiaSTTSampleRate.HZ_16000,
        base_url: str = DEFAULT_BASE_URL,
        options: CartesiaSTTOptions | None = None,
    ) -> None:
        if not api_key:
            raise ValueError("CartesiaSTT requires a non-empty api_key")

        if options is None:
            options = CartesiaSTTOptions(
                model=model,
                language=language,
                encoding=encoding,
                sample_rate=sample_rate,
            )

        self._api_key = api_key
        self._base_url = base_url
        self._opts = options

        self._session: aiohttp.ClientSession | None = None
        self._ws: aiohttp.ClientWebSocketResponse | None = None
        self._recv_task: asyncio.Task[None] | None = None
        self._keepalive_task: asyncio.Task[None] | None = None
        self._transcript_queue: asyncio.Queue[Transcript] = asyncio.Queue()
        self._running = False
        self.request_id: str | None = None
        self._audio_bytes_sent: int = 0

    @property
    def sample_rate(self) -> int:
        return self._opts.sample_rate

    @property
    def encoding(self) -> str:
        # Cartesia STT only accepts pcm_s16le today; surface a stable
        # observability label that maps to "linear16" semantics.
        return "linear16" if self._opts.encoding == "pcm_s16le" else self._opts.encoding

    def _record_transcript_cost(self) -> None:
        """Emit ``patter.cost.stt_seconds`` for buffered audio."""
        try:
            from getpatter.observability.attributes import record_patter_attrs

            bytes_per_sample = 1 if self.encoding == "mulaw" else 2
            seconds = self._audio_bytes_sent / float(
                self.sample_rate * bytes_per_sample
            )
            record_patter_attrs(
                {
                    "patter.cost.stt_seconds": seconds,
                    "patter.stt.provider": "cartesia",
                }
            )
            self._audio_bytes_sent = 0
        except Exception:  # pragma: no cover — defense in depth
            logger.debug("_record_transcript_cost failed", exc_info=True)

    def __repr__(self) -> str:
        return (
            f"CartesiaSTT(model={self._opts.model!r}, "
            f"language={self._opts.language!r}, sample_rate={self._opts.sample_rate})"
        )

    def _build_ws_url(self) -> str:
        """Construct the WebSocket URL with query parameters.

        Cartesia accepts auth via the ``api_key`` query param for the STT
        streaming endpoint.
        """
        base = self._base_url
        if base.startswith("http://"):
            base = "ws://" + base[len("http://") :]
        elif base.startswith("https://"):
            base = "wss://" + base[len("https://") :]
        elif not (base.startswith("ws://") or base.startswith("wss://")):
            base = "wss://" + base

        params = {
            "model": self._opts.model,
            "sample_rate": str(self._opts.sample_rate),
            "encoding": self._opts.encoding,
            "cartesia_version": API_VERSION,
            "api_key": self._api_key,
        }
        if self._opts.language:
            params["language"] = self._opts.language
        return f"{base}/stt/websocket?{urlencode(params)}"

    async def warmup(self) -> None:
        """Pre-call WebSocket warmup for the Cartesia STT ``/stt/websocket`` endpoint.

        Opens the WS (DNS + TLS + auth handshake), idles ~250 ms so the
        Cartesia edge keeps session state warm, then closes. By the time
        :meth:`connect` is invoked at call-pickup the resolver and TLS
        session are hot — net wire time saving of 200-500 ms.

        Billing safety: Cartesia STT bills on streamed audio seconds (per
        https://docs.cartesia.ai/2025-04-16/api-reference/stt/stt). Opening
        + closing the WebSocket without forwarding audio does not consume
        billable seconds. Best-effort: failures are logged at DEBUG.
        """
        ws_url = self._build_ws_url()
        headers = {"User-Agent": USER_AGENT}
        session: aiohttp.ClientSession | None = None
        ws: aiohttp.ClientWebSocketResponse | None = None
        try:
            session = aiohttp.ClientSession()
            ws = await asyncio.wait_for(
                session.ws_connect(ws_url, headers=headers),
                timeout=5.0,
            )
            # Idle briefly so the provider edge keeps session state warm.
            await asyncio.sleep(0.25)
        except aiohttp.WSServerHandshakeError as exc:
            # IMPORTANT: ``str(exc)`` includes the request URL, which
            # carries the API key as a query-string parameter (Cartesia
            # auth pattern). Log only the HTTP status so the API key
            # never lands in logs.
            logger.debug(
                "Cartesia STT warmup failed (best-effort): HTTP %d", exc.status
            )
        except Exception as exc:  # noqa: BLE001 - best-effort
            # The API key only travels in the URL, which only
            # ``WSServerHandshakeError`` exposes in ``str(exc)``. For
            # everything else (DNS, TCP, TLS, timeout) the exception
            # type alone is informative enough — and crucially never
            # leaks the URL.
            logger.debug(
                "Cartesia STT warmup failed (best-effort): %s",
                type(exc).__name__,
            )
        finally:
            if ws is not None:
                try:
                    await ws.close()
                except Exception:
                    pass
            if session is not None:
                try:
                    await session.close()
                except Exception:
                    pass

    async def connect(self) -> None:
        """Open the WebSocket and start recv + keepalive tasks."""
        if self._session is None:
            self._session = aiohttp.ClientSession()

        ws_url = self._build_ws_url()
        headers = {"User-Agent": USER_AGENT}
        self._ws = await self._session.ws_connect(ws_url, headers=headers)
        self._arm_recv_and_keepalive()

    async def open_parked_connection(
        self,
    ) -> tuple[aiohttp.ClientSession, aiohttp.ClientWebSocketResponse]:
        """Open a fresh WebSocket and return the session + WS without
        arming any recv / keepalive task.

        Used by :meth:`Patter._park_provider_connections` to park a
        Cartesia STT WS during the carrier ringing window so the per-call
        :class:`StreamHandler` can adopt it via :meth:`adopt_websocket`
        and skip the cold TLS + WS-upgrade handshake on the first turn.

        Billing safety: opening + parking the WS does not stream audio
        (Cartesia STT bills on streamed audio seconds), so no charge is
        incurred. Caller is responsible for closing both the WS and the
        session if the parked handle is never adopted.
        """
        session = aiohttp.ClientSession()
        try:
            ws_url = self._build_ws_url()
            headers = {"User-Agent": USER_AGENT}
            ws = await asyncio.wait_for(
                session.ws_connect(ws_url, headers=headers), timeout=10.0
            )
        except Exception:
            await session.close()
            raise
        return session, ws

    def adopt_websocket(
        self,
        session: aiohttp.ClientSession,
        ws: aiohttp.ClientWebSocketResponse,
    ) -> None:
        """Adopt a pre-opened, already-OPEN WebSocket parked by
        :meth:`open_parked_connection`. Skips the fresh WS handshake —
        audio frames can flow on the first turn instead of paying the
        ~150-400 ms TLS + WS-upgrade round-trip.

        Caller MUST verify the WS is still alive (``not ws.closed``)
        before calling. If the parked WS died between park and adopt,
        fall back to :meth:`connect`.
        """
        if self._session is None:
            self._session = session
        else:
            # Different session was already created (caller error /
            # connect was raced) — close the parked session to avoid
            # a leak.
            t = asyncio.create_task(session.close())
            t.add_done_callback(
                lambda f: (
                    not f.cancelled()
                    and f.exception()
                    and logger.debug(
                        "CartesiaSTT: leaked session close error: %s", f.exception()
                    )
                )
            )
        self._ws = ws
        self._arm_recv_and_keepalive()

    def _arm_recv_and_keepalive(self) -> None:
        """Start the receive + keepalive tasks against ``self._ws``.

        Shared between :meth:`connect` and :meth:`adopt_websocket` so
        the two paths produce byte-identical session state.
        """
        self._running = True
        self._recv_task = asyncio.create_task(self._recv_loop())
        self._keepalive_task = asyncio.create_task(self._keepalive_loop())

    async def send_audio(self, audio_chunk: bytes) -> None:
        """Forward a PCM s16le audio chunk to Cartesia."""
        if self._ws is None or self._ws.closed:
            raise RuntimeError("Not connected. Call connect() first.")
        self._audio_bytes_sent += len(audio_chunk)
        await self._ws.send_bytes(audio_chunk)

    async def finalize(self) -> None:
        """Force Cartesia to finalise the in-flight utterance immediately.

        Sends a ``finalize`` text frame on the live WebSocket. Cartesia
        replies with the final transcript followed by ``flush_done``,
        bypassing its conservative internal silence heuristic (which can
        wait 2-7 s on PSTN audio before naturally finalising). Wired
        into :meth:`StreamHandler` on the VAD ``speech_end`` event so
        the SDK's authoritative end-of-speech detection forces an
        immediate STT finalisation — turning Cartesia's natural-pause
        endpointing into a deterministic VAD-driven one, parity with
        the Deepgram fast-path. No-op when the WS isn't open.
        """
        if self._ws is None or self._ws.closed:
            return
        try:
            await self._ws.send_str(CartesiaSTTClientFrame.FINALIZE.value)
        except Exception as exc:  # noqa: BLE001 — defensive on a remote socket
            logger.debug("Cartesia finalize send failed: %s", exc)

    async def receive_transcripts(self) -> AsyncIterator[Transcript]:
        """Async generator yielding :class:`Transcript` events as they arrive."""
        while self._running or not self._transcript_queue.empty():
            try:
                transcript = await asyncio.wait_for(
                    self._transcript_queue.get(), timeout=0.1
                )
            except asyncio.TimeoutError:
                continue
            if transcript.is_final:
                self._record_transcript_cost()
            yield transcript

    async def _keepalive_loop(self) -> None:
        """Send WebSocket pings every 30 s to keep the session alive."""
        try:
            while self._running and self._ws is not None and not self._ws.closed:
                await asyncio.sleep(KEEPALIVE_INTERVAL_SECONDS)
                try:
                    await self._ws.ping()
                except Exception:  # noqa: BLE001
                    break
        except asyncio.CancelledError:
            return

    async def _recv_loop(self) -> None:
        """Read JSON frames from Cartesia and enqueue Transcripts."""
        assert self._ws is not None  # noqa: S101 — guaranteed by caller
        try:
            async for msg in self._ws:
                if msg.type == aiohttp.WSMsgType.TEXT:
                    try:
                        self._handle_event(json.loads(msg.data))
                    except Exception:  # noqa: BLE001
                        logger.exception("CartesiaSTT failed to process message")
                elif msg.type in (
                    aiohttp.WSMsgType.CLOSED,
                    aiohttp.WSMsgType.CLOSE,
                    aiohttp.WSMsgType.CLOSING,
                    aiohttp.WSMsgType.ERROR,
                ):
                    break
        finally:
            self._running = False

    def _handle_event(self, data: dict) -> None:
        """Parse a single Cartesia event and enqueue a Transcript if relevant.

        Protocol message types:
        - ``transcript``: interim (is_final=False) or final (is_final=True)
        - ``flush_done``: acknowledgement of a ``finalize`` request
        - ``done``: session is closing cleanly
        - ``error``: server-side error
        """
        message_type = data.get("type")
        if message_type == CartesiaSTTServerEvent.TRANSCRIPT:
            text = (data.get("text") or "").strip()
            is_final = bool(data.get("is_final", False))
            if not text and not is_final:
                return
            request_id = data.get("request_id")
            if request_id:
                self.request_id = request_id
            confidence = float(data.get("probability", 1.0) or 0.0)
            if text:
                self._transcript_queue.put_nowait(
                    Transcript(text=text, is_final=is_final, confidence=confidence)
                )
            return

        if message_type == CartesiaSTTServerEvent.ERROR:
            logger.error("Cartesia STT error: %s", data.get("message", "unknown"))
            return

        if message_type in (
            CartesiaSTTServerEvent.FLUSH_DONE,
            CartesiaSTTServerEvent.DONE,
        ):
            if message_type == CartesiaSTTServerEvent.DONE:
                self._running = False
            return

    async def close(self) -> None:
        """Send ``finalize``, close the WS, cancel tasks."""
        self._running = False
        if self._ws is not None and not self._ws.closed:
            try:
                await self._ws.send_str(CartesiaSTTClientFrame.FINALIZE.value)
            except Exception:  # noqa: BLE001
                pass
            await self._ws.close()
        self._ws = None

        for task in (self._recv_task, self._keepalive_task):
            if task is not None:
                task.cancel()
                try:
                    await task
                except (asyncio.CancelledError, Exception):  # noqa: BLE001
                    pass
        self._recv_task = None
        self._keepalive_task = None

        if self._session is not None:
            await self._session.close()
            self._session = None
